docs/features/nullable-reference-types.md

```diff
@@ -99,7 +99,24 @@ The `Interlocked.CompareExchange` methods have special handling in flow analysis
 
 When simple pre- and post-condition attributes are applied to a property, and an allowed input state is assigned to the property, the property's state is updated to be an allowed output state. For instance an `[AllowNull] string` property can be assigned `null` and still return not-null values.
 
-The use of any member returning `[MaybeNull]T` for an unconstrained type parameter `T` produces a warning, just like `default(T)` (see below). For instance, `listT.FirstOrDefault();` would produce a warning.
+
+Enforcing nullability attributes within method bodies:
+- An input parameter marked with [AllowNull] is initialized with a maybe-null (or maybe-default) state.
+- An input parameter marked with [DisallowNull] is initialized with a not-null state.
+- A parameter marked with [MaybeNull] or [MaybeNullWhen] can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with [NotNullWhen] (the attribute is ignored).
+- A parameter marked with [NotNull] will produce a warning when assigned a maybe-null value. Same for return values.
```

Comment:
This only applies to `out` and `ref` parameters right?

```suggestion
- A by-reference parameter marked with `[NotNull]` will produce a warning when assigned a maybe-null value. Same for return values.
```

Reply:
@sharwell, no this applies to by-value parameters as well. I'll update test `UnconditionalAttributesAffectMethodBodies_ReferenceType` to show that behavior.

---
In reply to: [373548582](https://github.com/dotnet/roslyn/pull/41336#discussion_r373548582) [](ancestors = 373548582)